File: backend/src/main/java/com/chominjungum/hangul/DictationCompare.java
package com.chominjungum.hangul;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Objects;

/**
 * 받아쓰기 채점 — Dart({@code hangul_core}) · TS({@code hangul-core-ts}) 구현과 동일 규칙.
 *
 * <p>서버는 학생이 보낸 점수를 믿지 않고 이 클래스로 <b>재채점</b>한다(PLAN §8.2).
 */
public final class DictationCompare {

    private DictationCompare() {
    }

    public record GlyphMatch(int index, boolean correct, String mismatchReason) {
        public Map<String, Object> toSummary() {
            Map<String, Object> map = new LinkedHashMap<>();
            map.put("i", index);
            map.put("ok", correct);
            if (mismatchReason != null) {
                map.put("why", mismatchReason);
            }
            return map;
        }
    }

    public record Result(
            int correctCount,
            int totalCount,
            double ratio,
            List<GlyphMatch> matches,
            String error,
            boolean hasExtraInput) {

        public int scorePercent() {
            return (int) Math.round(ratio * 100);
        }

        public List<Map<String, Object>> matchSummaries() {
            return matches.stream().map(GlyphMatch::toSummary).toList();
        }
    }

    public static Result score(String expected, String actual) {
        if (!HangulUtil.isHangul(expected)) {
            return new Result(0, 0, 0, List.of(), "정답이 jammin 허용 문자 집합이 아닙니다.", false);
        }
        if (!HangulUtil.isHangul(actual)) {
            return new Result(
                    0, HangulUtil.hangulSplit(expected).size(), 0, List.of(),
                    "답안이 jammin 허용 문자 집합이 아닙니다.", false);
        }

        List<Hangul> e = HangulUtil.hangulSplit(expected);
        List<Hangul> a = HangulUtil.hangulSplit(actual);
        List<GlyphMatch> matches = new ArrayList<>();
        int correct = 0;

        for (int i = 0; i < e.size(); i++) {
            Hangul exp = e.get(i);
            Hangul act = i < a.size() ? a.get(i) : null;
            boolean ok = glyphEquals(exp, act);
            if (ok) {
                correct++;
            }
            matches.add(new GlyphMatch(i, ok, ok ? null : reason(exp, act)));
        }

        for (int i = e.size(); i < a.size(); i++) {
            matches.add(new GlyphMatch(i, false, "불필요한 글자"));
        }

        int total = e.size();
        double ratio = total == 0 ? 1.0 : (double) correct / total;
        return new Result(correct, total, ratio, matches, null, a.size() > e.size());
    }

    private static boolean glyphEquals(Hangul e, Hangul a) {
        if (a == null) {
            return false;
        }
        if (!Objects.equals(e.getSpecialFlag(), a.getSpecialFlag())) {
            return false;
        }
        if (Boolean.TRUE.equals(e.getSpecialFlag())) {
            return Objects.equals(e.getSpecialType(), a.getSpecialType());
        }
        if (!Objects.equals(e.getErrorFlag(), a.getErrorFlag())) {
            return false;
        }
        return Objects.equals(e.getWord(), a.getWord())
                && e.getChoCode() == a.getChoCode()
                && e.getJungCode() == a.getJungCode()
                && e.getJongCode() == a.getJongCode();
    }

    private static String reason(Hangul e, Hangul a) {
        if (a == null) {
            return "글자 누락";
        }
        if (!Objects.equals(e.getSpecialFlag(), a.getSpecialFlag())) {
            return "문자 종류 불일치(한글/특수·숫자)";
        }
        if (Boolean.TRUE.equals(e.getSpecialFlag())) {
            return "기호·숫자 불일치: \"" + e.getSpecialType() + "\" vs \"" + a.getSpecialType() + "\"";
        }
        if (!Objects.equals(e.getWord(), a.getWord())) {
            return "글자 불일치: \"" + e.getWord() + "\" vs \"" + a.getWord() + "\"";
        }
        return "자모 불일치";
    }
}

File: packages/hangul-core-ts/src/dictation-compare.ts
/**
 * 받아쓰기 채점 — chominjungum Flutter 앱의 `DictationCompare`(Dart) 와 동일 규칙.
 *
 * 원본: chominjungum/packages/hangul_core/lib/src/dictation_compare.dart
 * 두 구현이 갈라지면 교실 모드(앱)와 온라인 모드(웹)의 점수가 달라진다.
 */

import { hangulSplit, isHangul, type HangulGlyph } from './hangul-util.js';

/** 한 글자 단위 비교 결과. */
export interface GlyphMatch {
  index: number;
  expected: HangulGlyph | null;
  actual: HangulGlyph | null;
  isCorrect: boolean;
  mismatchReason?: string;
}

export interface DictationScoreResult {
  correctCount: number;
  totalCount: number;
  ratio: number;
  matches: GlyphMatch[];
  error?: string;
  hasExtraInput: boolean;
}

export function scorePercent(result: DictationScoreResult): number {
  return Math.round(result.ratio * 100);
}

/** 서버 업싱크·저장용 축약형 (`attempt.glyph_matches`). */
export interface GlyphMatchSummary {
  i: number;
  ok: boolean;
  why?: string;
}

export function summarizeMatches(result: DictationScoreResult): GlyphMatchSummary[] {
  return result.matches.map((m) => ({
    i: m.index,
    ok: m.isCorrect,
    ...(m.mismatchReason ? { why: m.mismatchReason } : {}),
  }));
}

function glyphEquals(e: HangulGlyph, a: HangulGlyph | null): boolean {
  if (a === null) return false;
  if (e.specialFlag !== a.specialFlag) return false;
  if (e.specialFlag) {
    return e.specialType === a.specialType;
  }
  if (e.errorFlag !== a.errorFlag) return false;
  return (
    e.word === a.word &&
    e.choCode === a.choCode &&
    e.jungCode === a.jungCode &&
    e.jongCode === a.jongCode
  );
}

function reason(e: HangulGlyph, a: HangulGlyph | null): string {
  if (a === null) return '글자 누락';
  if (e.specialFlag !== a.specialFlag) {
    return '문자 종류 불일치(한글/특수·숫자)';
  }
  if (e.specialFlag) {
    return `기호·숫자 불일치: "${e.specialType}" vs "${a.specialType}"`;
  }
  if (e.word !== a.word) {
    return `글자 불일치: "${e.word}" vs "${a.word}"`;
  }
  return '자모 불일치';
}

/**
 * [expected] 정답, [actual] 학생 답(키보드 또는 OCR 정규화 텍스트).
 * 정답률은 **정답 글자 수 기준**이고, 초과 입력은 별도 match로 덧붙는다.
 */
export function score(expected: string, actual: string): DictationScoreResult {
  if (!isHangul(expected)) {
    return {
      correctCount: 0,
      totalCount: 0,
      ratio: 0,
      matches: [],
      error: '정답이 jammin 허용 문자 집합이 아닙니다.',
      hasExtraInput: false,
    };
  }
  if (!isHangul(actual)) {
    return {
      correctCount: 0,
      totalCount: hangulSplit(expected).length,
      ratio: 0,
      matches: [],
      error: '답안이 jammin 허용 문자 집합이 아닙니다.',
      hasExtraInput: false,
    };
  }

  const eGlyphs = hangulSplit(expected);
  const aGlyphs = hangulSplit(actual);
  const matches: GlyphMatch[] = [];
  let correct = 0;

  for (let i = 0; i < eGlyphs.length; i++) {
    const e = eGlyphs[i]!;
    const a = i < aGlyphs.length ? aGlyphs[i]! : null;
    const ok = glyphEquals(e, a);
    if (ok) correct++;
    matches.push({
      index: i,
      expected: e,
      actual: a,
      isCorrect: ok,
      ...(ok ? {} : { mismatchReason: reason(e, a) }),
    });
  }

  for (let i = eGlyphs.length; i < aGlyphs.length; i++) {
    matches.push({
      index: i,
      expected: null,
      actual: aGlyphs[i]!,
      isCorrect: false,
      mismatchReason: '불필요한 글자',
    });
  }

  const total = eGlyphs.length;
  return {
    correctCount: correct,
    totalCount: total,
    ratio: total === 0 ? 1 : correct / total,
    matches,
    hasExtraInput: aGlyphs.length > eGlyphs.length,
  };
}
